aten/src/ATen/core/ivalue_inl.h

```diff
@@ -468,11 +468,21 @@ struct C10_EXPORT ivalue::Future : c10::intrusive_ptr_target {
     auto fut = createInstance(std::move(type));
     addCallback(
         [fut, cb = std::move(callback)]() {
+          // Get the true stacktrace if call to markCompleted() fails. We cannot
+          // simply call fut->markCompleted(cb()) since if markCompleted fails
+          // after setting completed_ flag, setError will report error that
+          // completed_ is set, instead of the true error.
+          at::IValue cbValue;
+          bool hasException = false;
           try {
-            fut->markCompleted(cb());
+            cbValue= cb();
           } catch (std::exception&) {
+            hasException = true;
             fut->setError(std::current_exception());
```

Comment:
Shouldn't calling `setErrorIfNeeded` here be sufficient? It would log the appropriate exception and wouldn't result in the weird assertion error.

Reply:
@lw's comment here https://github.com/pytorch/pytorch/issues/53717#issuecomment-800545655 says why that wouldn't work, in particular:

> If we just replaced setError with setErrorIfNeeded in then the end result would be that the future is still marked as successfully completed, whereas the CUDA part isn't set-up correctly, which could lead to silent CUDA synchronization/correctness issues.

Reply:
Maybe let's add a comment in the code for why we can't use `setErrorIfNeeded`, to warn the next person that comes along and wonders the same thing?

Reply:
I see, would a simpler fix be to set `completed_ = true` after `postMarkCompletedHook`? If we don't want to mark a future complete without running the `postMarkCompletedHook` successfully, then probably we should move `completed_ = true` accordingly?